## Embryo Options - Inventory Exercises - Data Analyst
### Linnea Miller

In [1]:
import pandas as pd

In [2]:
# Read in .csv files
charges = pd.read_csv('./PostCharges - Export.csv', parse_dates=[3])
territories = pd.read_csv('./StateProvinceTerritories.csv')
countries = pd.read_csv('./Countries.csv')

In [3]:
charges.head(3)

EMR   Last Name   First Name        DOB  \
0  503113.0  Last503113  First503113 1972-11-04   
1  909895.0  Last909895  First909895 1987-11-20   
2   93361.0   Last93361   First93361 1990-05-13   

                                 Partner # Embryos  # Oocytes Tank  \
0  PartnerLast503113, PartnerFirst503113         3        0.0    V   
1  PartnerLast909895, PartnerFirst909895         3        0.0    S   
2    PartnerLast93361, PartnerFirst93361         6        0.0    J   

  Canister-Color Donated From            ...             Billing Month  \
0        11-Blue          NaN            ...                  December   
1       7-Silver          NaN            ...                  December   
2         8-Pink          NaN            ...                      July   

  Comments     Address 1 Address 2            City State    Zip Country  \
0      NaN  107 Main St.       NaN     West Jordan    UT  84081     USA   
1      NaN  863 Main St.       NaN  Salt Lake City    UT  84106     USA   
2      NaN  216 Main St.       NaN          Moroni    UT  84646     USA   

  Primary Phone                      Email  
0  3.054591e+09  notanemail884@example.com  
1  6.186166e+09   notanemail50@example.com  
2  6.344384e+09  notanemail512@example.com  

[3 rows x 21 columns]

In [4]:
charges.columns

Index(['EMR', 'Last Name', 'First Name', 'DOB', 'Partner', '# Embryos',
       '# Oocytes', 'Tank', 'Canister-Color', 'Donated From', 'Date In',
       'Billing Month', 'Comments', 'Address 1', 'Address 2', 'City', 'State',
       'Zip', 'Country', 'Primary Phone', 'Email'],
      dtype='object')

In [5]:
territories.head(3)

ID Country Abbreviation State Abbreviation        State Name
0  AT1                   AT                  1        Burgenland
1  AT2                   AT                  2           Karnten
2  AT3                   AT                  3  Niederosterreich

In [6]:
countries.head(3)

Country Abbreviation  \
0                   AF   
1                   AL   
2                   DZ   

  Country Name (*This sheet should be sorted by this column for LOOKUP function to work)  
0                                        Afghanistan                                      
1                                            Albania                                      
2                                            Algeria

### Question 1

*Q: How many unique patients are there in the dataset?*

A: 2201 unique patients

In [7]:
# Look at shape of the table to see TOTAL # of rows
charges.shape

(2201, 21)

In [8]:
# Using set function to find unique rows (i.e. unique patients)
len(set(charges['EMR']))

2201

In [9]:
# charges['Last Name'].describe()

In [10]:
# charges[charges['Last Name'] == 'Last915553']

### Question 2

*Q: Identify all patients without an EMR number.*

A: Find all patients where EMR is null.

In [11]:
charges[charges['EMR'].isnull()]

EMR   Last Name   First Name        DOB  \
17    NaN  Last915553     First212 1984-07-12   
33    NaN   Last26280  First292222 1994-01-19   
55    NaN  Last260308  First982213 1981-04-26   
111   NaN  Last802195  First866153 1988-06-30   
293   NaN  Last579476   First31068 1985-10-27   
328   NaN  Last880108  First302075 1981-05-04   
352   NaN  Last356408  First886705 1984-12-15   
396   NaN   Last96756  First405875 1974-12-13   
436   NaN  Last962634  First924782 1985-03-15   
439   NaN   Last61463  First747613 1990-07-24   
465   NaN  Last859958  First993365 1985-08-27   
639   NaN  Last324604  First355242 1981-11-29   
666   NaN  Last381834  First569583 1988-03-22   
723   NaN  Last369517  First373172 1978-01-21   
759   NaN  Last469511  First297765 1986-12-19   
768   NaN  Last627528  First864783 1973-10-13   
781   NaN  Last953433  First506886 1979-04-28   
811   NaN  Last580108  First592377 1984-03-20   
820   NaN  Last870979  First672799 1980-10-30   
834   NaN  Last222782  First228225 1986-10-17   
870   NaN  Last597782   First73148 1985-10-22   
930   NaN  Last572846  First899854 1975-05-09   
955   NaN  Last823010  First944212 1973-04-27   
959   NaN  Last643553  First696192 1985-11-08   
1022  NaN  Last735826  First628619 1973-04-05   
1077  NaN  Last840357  First770055 1972-01-29   
1088  NaN   Last47569  First808181 1972-08-29   
1146  NaN  Last845197  First993839 2067-02-20   
1160  NaN  Last878313  First799908 1971-05-02   
1169  NaN   Last48904  First985364 1979-10-07   
...   ...         ...          ...        ...   
1237  NaN  Last248432  First940452 1982-12-28   
1304  NaN  Last946835  First976583 1973-01-26   
1307  NaN  Last122668  First264421 1982-11-29   
1333  NaN  Last553573  First920413 1982-08-20   
1356  NaN  Last563854  First932133 1975-09-10   
1397  NaN  Last786702  First475246 1978-12-26   
1408  NaN  Last923953  First535158 1991-02-15   
1450  NaN  Last103700  First461174 1981-08-10   
1516  NaN  Last453578  First119959 1977-10-11   
1540  NaN  Last148790  First386798 1988-06-18   
1574  NaN  Last486071  First330846 1980-12-15   
1660  NaN  Last361877  First109848 1981-04-27   
1662  NaN  Last679610  First907772 1982-06-19   
1697  NaN  Last766203  First913822 1972-01-24   
1799  NaN  Last588040  First355234 1986-04-23   
1807  NaN  Last908150  First532707 1979-05-12   
1841  NaN  Last181483  First551960 1990-09-26   
1923  NaN  Last268064  First423067 1982-04-22   
1939  NaN  Last431020  First533270 1981-05-04   
1952  NaN  Last491951  First606614 1990-03-02   
1960  NaN  Last812536  First817862 1989-05-15   
1992  NaN  Last233625  First716170 1988-04-18   
2043  NaN  Last309176  First872381 1986-09-05   
2061  NaN  Last297657  First732294 1977-05-20   
2108  NaN  Last119935  First324169 1987-11-17   
2118  NaN  Last611451  First984179 1977-05-12   
2127  NaN  Last270088  First576505 1978-04-26   
2157  NaN  Last137427  First775773 1985-06-27   
2178  NaN  Last981743  First666242 1988-06-14   
2189  NaN  Last246988  First419376 1980-08-26   

                                    Partner # Embryos  # Oocytes  \
17    PartnerLast804628, PartnerFirst748242         3        0.0   
33    PartnerLast895676, PartnerFirst384420         4        0.0   
55    PartnerLast294652, PartnerFirst328208         3        0.0   
111   PartnerLast111122, PartnerFirst888945         8        0.0   
293   PartnerLast152672, PartnerFirst647124         1        0.0   
328   PartnerLast276505, PartnerFirst184089         4        0.0   
352    PartnerLast294027, PartnerFirst55172         1        0.0   
396                                     NaN         2        4.0   
436   PartnerLast302823, PartnerFirst371216         5        0.0   
439   PartnerLast548882, PartnerFirst146467         5        0.0   
465   PartnerLast685237, PartnerFirst269213         6        0.0   
639   PartnerLast639758, PartnerFirst658491         7        0.0   
666    PartnerLast73264, PartnerFirst978000         3        0.0   
723   PartnerLa

### Question 3

*Q: Identify all patients with an invalid Date of Birth.*

A: When changing the datatype from string to datetime, any row with a DOB *before the year of 1969* get's incorrectly converted to future dates that don't exist yet as birthdays. 

For example, 11/25/68 becomes **2068**-11-25, whereas 11/20/69 becomes **1969**-11-20

In [12]:
# # Change from string to datetime format
# charges['DOB'] = pd.to_datetime(charges['DOB'])

In [13]:
# charges['DOB'].sort_values().tail(50)

Before 1969, the dates are being changes to invalid dates: 2067, 2068, etc...

Notice that the dates in 2019 are fine, but then starting with 5/11/55, the year is converted incorrectly

In [14]:
charges.sort_values(by='DOB').tail(10)

EMR   Last Name   First Name        DOB  \
1439  882522.0  Last882522  First882522 2068-03-07   
1758  859294.0  Last859294  First859294 2068-04-08   
867    79040.0   Last79040   First79040 2068-04-16   
882   535845.0  Last535845  First535845 2068-04-29   
866     7619.0    Last7619    First7619 2068-06-30   
2004  455712.0  Last455712  First455712 2068-06-30   
1526  299449.0  Last299449  First299449 2068-10-15   
1597  450371.0  Last450371  First450371 2068-11-02   
1212  528499.0  Last528499  First528499 2068-11-21   
1269  540937.0  Last540937  First540937 2068-11-25   

                                    Partner # Embryos  # Oocytes  \
1439  PartnerLast882522, PartnerFirst882522         1        0.0   
1758  PartnerLast859294, PartnerFirst859294         7       10.0   
867     PartnerLast79040, PartnerFirst79040         5        0.0   
882   PartnerLast535845, PartnerFirst535845        19        0.0   
866       PartnerLast7619, PartnerFirst7619         2        0.0   
2004  PartnerLast455712, PartnerFirst455712         6        0.0   
1526                                    NaN        17        0.0   
1597  PartnerLast450371, PartnerFirst450371         9        0.0   
1212                                    NaN         9        0.0   
1269                                    NaN         7        0.0   

                                                Tank  \
1439                                               V   
1758  N                  E                         N   
867                                                S   
882                                                S   
866                                                G   
2004                                               F   
1526                                               S   
1597                                               V   
1212                                               V   
1269                                               S   

                                         Canister-Color Donated From  \
1439                                              8-Red          NaN   
1758  2-Yellow                 10-Yellow            ...          NaN   
867                                            3-Silver          NaN   
882                                            5-Yellow          NaN   
866                                           10-Purple          NaN   
2004                                             1-Blue          NaN   
1526                                           3-Yellow          NaN   
1597                                           7-Yellow          NaN   
1212                                          11-Orange          NaN   
1269                                           9-Silver          NaN   

                ...             Billing Month     Comments     Address 1  \
1439            ...                   October          NaN  821 Main St.   
1758            ...                 September          NaN  234 Main St.   
867             ...                      June  Ext. Freeze  116 Main St.   
882             ...                      June          NaN  729 Main St.   
866             ...                    August          NaN           NaN   
2004            ...                     April  Ext. Freeze  896 Main St.   
1526            ...                 September          NaN  322 Main St.   
1597            ...                   October          NaN           NaN   
1212            ...                   January          NaN           NaN   
1269            ...                  February          NaN           NaN   

     Address 2              City State    Zip Country Primary Phone  \
1439       NaN           Chicago    IL  60607     USA  6.019750e+09   
1758       NaN               NaN   NaN    NaN     NaN  5.812522e+09   
867        NaN  Saint Petersburg    FL  33715     USA  4.967012e+08   
882        NaN          Woodside    CA  94062     USA  8.017665e+09   
866        NaN               NaN   NaN    NaN     NaN           N

### Question 4

*Q: Identify all patient addresses that do not have a valid country or state code.*

A:

1. Isolate rows where `Address 1` is populated.

In [15]:
has_address = charges[['Address 1', 'Address 2', 'City', 'State', 'Zip', 'Country']][charges['Address 1'].notnull()]

# Slice of what table currently looks like - There are still rows 
has_address.loc[2170:2177]

Address 1 Address 2         City State    Zip Country
2170  544 Main St.       NaN        Provo    UT  84601     USA
2171  361 Main St.       NaN   Washington    UT  84780     USA
2174  502 Main St.       NaN  West Jordan    UT  84081     USA
2175  127 Main St.       NaN          NaN   NaN    NaN     NaN
2176  276 Main St.       NaN       Lindon    UT  84042     USA
2177  966 Main St.       NaN      Midvale    UT  84047     USA

In [23]:
has_address

Address 1 Address 2             City State    Zip Country
0     107 Main St.       NaN      West Jordan    UT  84081     USA
1     863 Main St.       NaN   Salt Lake City    UT  84106     USA
2     216 Main St.       NaN           Moroni    UT  84646     USA
3     932 Main St.       NaN        Bluffdale    UT  84065     USA
4       4 Main St.       NaN   Eagle Mountain    UT  84005     USA
5     563 Main St.       NaN       Cedar City    UT  84720     USA
6     131 Main St.       NaN  North Salt Lake    UT  84054     USA
7     210 Main St.       NaN    American Fork    UT  84003     USA
8     573 Main St.       NaN     Rock Springs    WY  82601     USA
9     926 Main St.       NaN        Arlington    VA  22206     USA
10     24 Main St.       NaN   Eagle Mountain    UT  84005     USA
11    358 Main St.       NaN      West Jordan    UT  84081     USA
12    319 Main St.       NaN      West Valley    UT  84119     USA
13    171 Main St.       NaN    San Francisco    CA  94107     USA
15    449 Main St.       NaN             Orem    UT  84057     USA
16    549 Main St.       NaN      Los Angeles    CA  90045     USA
17    453 Main St.       NaN             Lehi    UT  84043     USA
18    352 Main St.       NaN            Provo    UT  84606     USA
19    584 Main St.       NaN      North Logan    UT  84341     USA
20    917 Main St.       NaN         Highland    UT  84003     USA
21    504 Main St.       NaN        Pocatello    ID  83201     USA
22    807 Main St.       NaN           Lindon    UT  84042     USA
23    214 Main St.       NaN     South Jordan    UT  84095     USA
24    180 Main St.       NaN     Taylorsville    UT  84129     USA
25    653 Main St.       NaN            Boise    ID  83713     USA
26    439 Main St.       NaN        St George    UT  84790     USA
27    812 Main St.       NaN      West Jordan    UT  84084     USA
28    318 Main St.       NaN      Springville    UT  84663     USA
29    677 Main St.       NaN        Henderson    NV  89014     USA
30    202 Main St.       NaN         Herriman    UT  84096     USA
...            ...       ...              ...   ...    ...     ...
2164   16 Main St.       NaN        Kaysville    UT  84037     USA
2165  914 Main St.       NaN      Cedar Hills    UT  84062     USA
2166   67 Main St.       NaN   Salt Lake City    UT  84106     USA
2167  612 Main St.       NaN      Idaho Falls    ID  83401     USA
2168  631 Main St.       NaN      Idaho Falls    ID  83401     USA
2169  634 Main St.       NaN     Spanish Fork    UT  84660     USA
2170  544 Main St.       NaN            Provo    UT  84601     USA
2171  361 Main St.       NaN       Washington    UT  84780     USA
2174  502 Main St.       NaN      West Jordan    UT  84081     USA
2175  127 Main St.       NaN              NaN   NaN    NaN     NaN
2176  276 Main St.       NaN           Lindon    UT  84042     USA
2177  966 Main St.       NaN          Midvale    UT  84047     USA
2178  335 Main St.       NaN       Glen Allen    VA  23060     USA
2179  257 Main St.       NaN             Orem    UT  84057     USA
2181  373 Main St.       NaN   Pleasant Grove    UT  84062     USA
2182  786 Main St.       NaN      Idaho Falls    ID  83404     USA
2183  162 Main St.       NaN          Lapoint    UT  84039     USA
2184  757 Main St.       NaN     South Jordan    UT  84009     USA
2187  177 Main St.       NaN             Elko    NV  89801     USA
2188  269 Main St.       NaN      Los Angeles    CA  90027     USA
2189  276 Main St.       NaN           Payson    UT  84651     USA
2190   27 Main St.       NaN  North Salt Lake    UT  84054     USA
2191   68 Main St.       NaN             Reno    NV  89521     USA
2192  941 Main St.       NaN          Rexburg    ID  83440     USA
2193  108 Main St.       NaN             Reno    NV  89501     USA
2195  128 Main St.       NaN   Salt Lake City    UT  84121     USA
2197   69 Main St.       NaN            Sandy    UT  84070     USA
2198  601 Main St.       NaN           Vernal    XY  840

2. Compare country/state codes in `charges` table with those in the `countries` and `territories` tables.

In [16]:
countries = list(countries['Country Abbreviation'])

In [18]:
cities = list(territories['State Abbreviation'])

3. Drill down further where `State` or `Country` is **NOT** populated or populated **incorrectly**.

In [21]:
has_address[(has_address['State'].isnull()) | (has_address['Country'].isnull()) | (has_address['Country'].isin(countries))]

Address 1            Address 2       City State  Zip Country
218   173 Main St.                  NaN    Alberta   NaN  NaN      CA
1720  731 Main St.  Seuzach Switzerland        NaN   NaN  NaN     NaN
1758  234 Main St.                  NaN        NaN   NaN  NaN     NaN
1812  992 Main St.                  NaN  Vancouver    BC   CA     NaN
2175  127 Main St.                  NaN        NaN   NaN  NaN     NaN

In [22]:
has_address[has_address['Country'].isin(countries)]

Address 1            Address 2       City State  Zip Country
218   173 Main St.                  NaN    Alberta   NaN  NaN      CA
1720  731 Main St.  Seuzach Switzerland        NaN   NaN  NaN     NaN
1758  234 Main St.                  NaN        NaN   NaN  NaN     NaN
1812  992 Main St.                  NaN  Vancouver    BC   CA     NaN
2175  127 Main St.                  NaN        NaN   NaN  NaN     NaN

##### Observations

- Notice `has_address.iloc[1720]` which has city/state/country information in the `Address 2` column instead of in the *appropriate* columns. Based on the `countries` table, the country code for Switzerland would be `CH`.
- `has_address.iloc[1812]` has an error. The country code appears in the `'Zip'` column instead of the `'Country'` column. If `'CA'` was in the `'Country'` column, it *would* be a valid country code!!


In [ ]:
# Rename column to simplify
countries = countries.rename(columns={'Country Name (*This sheet should be sorted by this column for LOOKUP function to work)':'Country Name'})

In [ ]:
# Verify Switzerland is valid country name & find out what the country abbreviation is for this specific row
countries[countries['Country Name'] == 'Switzerland']

In [ ]:
# Verify CA is valid country abbreviation
countries[countries['Country Abbreviation'] == 'CA']

##### 2. `State codes`

In [ ]:
states = list(set(list(charges['State'])))
states.pop(0) # Remove null from list

In [ ]:
# territories.head()
# states

In [ ]:
territories[territories['State Abbreviation'].isin(states)]

In [ ]:
# Number of rows where 'State' is populated
len(charges[charges['State'].isin(states)])

##### Nulls

In [ ]:
null_1 = list(charges[charges['State'].isnull() & charges['Country'].isnull()].index)
len(null_1)

There are 141 rows where *both* the State Code and Country Code are null.

In [ ]:
null_2 = list(charges[charges['State'].isnull() & charges['Country'].notnull()].index)
len(null_2)

There are 9 rows where *only* the State Code is null.

In [ ]:
null_3 = list(charges[charges['State'].notnull() & charges['Country'].isnull()].index)
len(null_3)

There is 1 row where *only* the Country is null.

In [ ]:
null_3

In [ ]:
# Dig into this specific case - where the Country is null but there's a city and state
charges.iloc[null_3]

In [ ]:
# # Get list of index values for all nulls
# all_nulls = null_1 + null_2 + null_3

# # See all rows that contain nulls for 'City' and 'Country'
# charges[['State', 'Country']].iloc[all_nulls]

### Question 5

*Q: Identify all records where the 'Billing Month' does not match the month of the 'Date In'.*

A: Create two new columns that contain the numeric value of the months from both original columns (*'Billing Month'* and *'Date In'*). Do this by converting each column to datetime in some form and pulling the **month** information. Finally, find the rows where the values in the two newly created columns *don't* match.

In [ ]:
from datetime import datetime

In [ ]:
# Check for null values
charges[['Date In', 'Billing Month']].isnull().sum()

In [ ]:
# Replace all null cells with a string ('?') and leave all other cells as they are
charges['Billing Month'] = charges['Billing Month'].map(lambda x: '?' if str(x) == 'nan' else str(x))
charges['Date In'] = charges['Date In'].map(lambda x: '?' if str(x) == 'nan' else x)

In [ ]:
# Create function that will yield the corresponding number to the 'Billing Month'
def month1(x):
    d = datetime.strptime(x, '%B')
    if x != '?':
        return int(d.strftime('%m'))
    else:
        return x

In [ ]:
# Create function that will yield the corresponding number to the 'Date In'
def month2(x):
    try: # Using try/except here in case there are any weird strings that don't fit the format I define below
        d = datetime.strptime(x, '%m/%d/%y')
        if x != '?':
            return int(d.strftime('%m'))
    except:
        return x

In [ ]:
# Create new column with 'Billing Month' converted into corresponding number (int) if doesn't equal '?'
# Using month1 function from above
charges['Bill_Month_num'] = charges['Billing Month'].map(lambda x: month1(x) if x != '?' else x)

In [ ]:
# Create new column with 'Date In' converted into corresponding number (int) if doesn't equal '?'
# Using month2 function from above
charges['Date_In_num'] = charges['Date In'].map(lambda x: month2(x) if x != '?' else x)

In [ ]:
# First 30 rows - noticing that there are cells in the Date_In_num column that did not get converted because there is >1 date in the string
charges[['Bill_Month_num', 'Date_In_num']].head(30)

In [ ]:
# Find the rows where the month values don't match
non_match = charges[['Bill_Month_num', 'Date_In_num']][charges['Bill_Month_num'] != charges['Date_In_num']]

In [ ]:
# Find how many rows out of the whole dataset have mismatching months
len(non_match)

In [ ]:
# Show records where 'Billing Month' does not match the month of the 'Date In'
non_match.head(10)
# non_match

After cleaning the data to an extent, I've found two patterns concerning the rows containing months that "don't match":

1. `'Bill_Month_num'` = `'?'` when `'Date_In_num'` is some integer (**ex**: `charges.iloc[26]`)
2. `'Bill_Month_num'` is some integer when `'Date_In_num'` remains a string that contains *two* dates (**ex**: `charges.iloc[43]`)

That being said, you could clean/manipulate the `'Date_In_num'` values even further to separate the multiple date situations and write a function to see if the month in *either* date matched the `'Bill_Month_num'` value in each case. 

It would obviously take additional research to figure out why some of the `'Bill_Month_num'` rows contained nulls (i.e. '?').

### Question 6

*Q: Create a combined list of all non-zero embryo and egg records (so we can track embryo and egg records separately.*

A:

### Question 7

### Question 8

##### Invalid Codes

##### 1. `Country codes`

In [ ]:
# Look at all possible codes in the countries dataset
country_codes = list(set(list(charges['Country'])))
country_codes.pop(0) # Remove null from list

In [ ]:
country_codes

In [ ]:
countries.head(3)

In [ ]:
# Look up Country Abbreviation values  from abovein countries table
countries[countries['Country Abbreviation'].isin(country_codes)]

Looking up the 3 country codes that aren't null we have in our `charges` table, only `'CA'` can be confirmed as a **valid** `Country Abbreviation`. We have to dig deeper into the other country codes.

In [ ]:
# Look up Country Abbreviation values in countries table
countries[countries['Country Abbreviation'].isin(['US', 'USA'])]

The rows in `charges` where `Country` == `'USA'` have an invalid country code. The official country abbreviation is `'US'`, not `'USA'`.

In [ ]:
# Look up Country Abbreviation values in countries table
countries[countries['Country Abbreviation'].isin(['ES', 'ESP', 'EP', 'SP'])]

The rows in `charges` where `Country` == `'ESP'` have an invalid country code. The official country abbreviation is `'ES'`, not `'ESP'`.